### PR TITLE
monitoring: add INJECT_LABEL_MATCHERS for injecting matchers on all selectors

### PR DESCRIPTION
**Pull request by bobheadxi**
Allows injection of arbitrary label matchers (`INJECT_LABEL_MATCHERS`) into all selectors in Prometheus queries within the monitoring generator (`metric_name` or `metric_name{foo="bar"}`) by:

1. replacing all template variables in queries, such that they become valid queries (introduced in https://github.com/sourcegraph/sourcegraph/pull/41643)
2. parsing all queries into PromQL abstract syntax trees
3. traversing the PromQL AST for `VectorSelector`s and injecting the desired label matchers
4. rendering the modified AST back into a query string
5. reversing the template variable replacement so that the query contains the desired template variables again

For alerts, we hardcode the desired label into the alert rules, since we cannot allow records to have arbitrary labels. This new processing also formats queries when rendered into docs/dashboards/etc.

Additional patches are required to pull parts of the dashboard and rules definition into the generator so that we can apply this processing universally:

- https://github.com/sourcegraph/sourcegraph/pull/41769
- https://github.com/sourcegraph/sourcegraph/pull/41712
- https://github.com/sourcegraph/sourcegraph/pull/41707
- https://github.com/sourcegraph/sourcegraph/pull/41705

The goal of this functionality is that if we can set up all metrics from a particular customer to export with a label, e.g. `customer=foobar`, we can generate and upload dashboards into folders that select on `customer=` labels, allowing us to aggregate all instance metrics on a single instance with dashboards that select for each separate instance.

Related: https://github.com/sourcegraph/sourcegraph/issues/35380, https://github.com/sourcegraph/customer/issues/1151

## Test plan

`INJECT_LABEL_MATCHERS='foo=bar' go generate ./monitoring/...` => https://github.com/sourcegraph/sourcegraph/commit/9865e902bf9f89416fa9fba94975e7f867fd21c5

```diff
- Query: `histogram_quantile(0.99, sum by (le)(rate(src_search_streaming_latency_seconds_bucket{source="browser"}[5m])))`
+ Query: `histogram_quantile(0.99, sum by(le) (rate(src_search_streaming_latency_seconds_bucket{foo="bar",source="browser"}[5m])))`
```

- [x] Test dashboards with `ns="prod"` label in dotcom
- [x] validate that we can upload dashboards to folders => **this just requires setting up a custom provisioning YAML a la [this config](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1c05f88ac5594174d26adb15e6341024b78b252a/-/blob/docker-images/grafana/config/provisioning/dashboards/sourcegraph.yaml?L16:1=) which lets us send dashboards to various folders**
- [x] Assess risk of template variable substitution and substitution reversal - the reversal might pick up values that were not part of the replacement in the first place => **numeric example values are now scrambled somewhat**
- [x] Ensure all queries that need to be modified are modified in the preprocessing: https://github.com/sourcegraph/sourcegraph/pull/41707, https://github.com/sourcegraph/sourcegraph/pull/41712
- [x] Manual validation of rendered queries

Dashboard testing:

```yaml
commands:
  prometheus:
    # install can just be set up gcloud credentials for a cluster
    # e.g. https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/deployments/instances
    install: gcloud container clusters get-credentials cloud --zone us-central1-f --project sourcegraph-dev
    # make remote prometheus accessible to local grafana
    cmd: kubectl port-forward -n prod svc/prometheus 9090:30090

  monitoring-generator:
    env:
      # don't reload your production Prometheus!
      PROMETHEUS_DIR: ''
      INJECT_LABEL_MATCHERS: 'ns=prod'
```

then `sg start monitoring`

A new, simple e2e test to ensure this doesn't error out is being introduced further up the stack: https://github.com/sourcegraph/sourcegraph/pull/41766

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #41642** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41642" target="_blank">Graphite</a> 
    * **PR #41643** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41643" target="_blank">Graphite</a> 
      * **PR #41644** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41644" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41644?utm_source=stack-comment).

**Review by jhchabran — APPROVED** [review]
Nice!

File: monitoring/monitoring/internal/promql/promql.go [hunk 1]
```diff
@@ -11,6 +12,38 @@ func Validate(expression string, vars VariableApplier) error {
 	return err
 }
 
+func Inject(expression string, matchers []*labels.Matcher, vars VariableApplier) (string, error) {
```
Comment (danieldides) [inline monitoring/monitoring/internal/promql/promql.go:15]: nit: can we add a docstring since this is an exported function?

**Review by danieldides — APPROVED** [review]
All looks good to me!

File: monitoring/monitoring/internal/promql/promql.go (hunk 1 above)
Comment (bobheadxi) [reply monitoring/monitoring/internal/promql/promql.go:20]: Good call out, I've expanded on everything in this package with some additional docstrings :)

**Comment by stefanhengl** [comment]
@bobheadxi 

@ggilmore and I noticed that the labels of the Zoekt dashboard are broken and used git bisect to trace it back to this commit. Any idea what is going on?

Details:
The legend shows twice "tracked", but one of the labels should be "assigned".
image

**Comment by bobheadxi** [comment]
@stefanhengl sorry for the late response, I just came across this notification - in the future please feel free to ping me in Slack directly to take a look at something if I miss it!

looking at that panel configuration if looks like some custom options are used to hijack the targets to enable multiple queries on a panel - this is not something we support. Especially with the recent changes to leverage `Query` to allowlist metrics we collect on Cloud, I'd encourage that this panel be split up because they are tracking different metrics. You can also use a query on `__name__` like so: https://github.com/sourcegraph/sourcegraph/pull/44621

**Pull request merged**